pkg/kgo/consumer_group.go

```diff
@@ -1103,6 +1103,12 @@ func (g *groupConsumer) handleJoinResp(resp *kmsg.JoinGroupResponse) (restart bo
 	}
 
 	leader := resp.LeaderID == resp.MemberID
+	if *g.cfg.instanceID != "" {
+		leader = strings.HasPrefix(resp.LeaderID, *g.cfg.instanceID)
```

Comment:
@twmb Hi Travis, yes, since we are using the static membership and found when we update the topic partitions, there is no leader rejoining to update the metadata and then we found:
```
kafkaproxy.leader | false
-- | --
  | kafkaproxy.leader_id | 1bd690c4-a8d3-4abb-a7d8-1b4ba71fef75-8cbcafb1-36b3-45bf-b583-1101098a41fc
  | kafkaproxy.lvlno | 20
  | kafkaproxy.member_id | 1bd690c4-a8d3-4abb-a7d8-1b4ba71fef75-99dfebc4-b602-4a7a-b7d3-c5afe9e35c82

```
when we rotate/replace the same group instance id with a new member, although the join response shows the previous group instance id holder is leader but the new member with a new UUID from broker side in this condition `leader := resp.LeaderID == resp.MemberID` is still false which resulting although it is the "real leader" but the `g.leader` was set false and `findNewAssignments` in the `doOnMetadataUpdate` callback failed to trigger rejoin.

Reply:
I don't think the above is correct. This isn't described in KIP-345, and I just checked the Kafka source:

https://github.com/apache/kafka/blob/d0ff8697189bda65b5381a95cccf6c4428b08637/clients/src/main/java/org/apache/kafka/common/requests/JoinGroupResponse.java#L42

isLeader is only if the leader ID matches the member ID exactly.

With static assignment, when the member with the same instance ID joins, it receives its old assignment.

I think the more problematic code is in findNewAssignments, specifically right here: https://github.com/twmb/franz-go/blob/93876346fa2f08457befb04dd5eff320b6a48e1b/pkg/kgo/consumer_group.go#L1619 (although the leader.get a few lines up looks suspicious as well in light of this investigation). The kgo client does not send a JoinGroup request if the client is not a leader. This isn't correct in the context of instance IDs. The Kafka core code supports replacing the member ID for static members: https://github.com/apache/kafka/blob/d0ff8697189bda65b5381a95cccf6c4428b08637/core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala#L331-L333

I'll remove the `leader.get` aspect in the client. That should fix things on your side.

Reply:
This is slightly more problematic than I thought. Removing the leader.get does not automatically fix this, there is one extra needed step. I've re-read KIP-345 a few times just to be sure (https://cwiki.apache.org/confluence/display/KAFKA/KIP-345).

The key lines:
```
Update 2/12: It turns out that we could cover the leader rejoin case for static membership. When the leader joins with non-empty member.id, it indicates that the leader is joining with a different purpose than rolling bounce. If leader joins with UNKNOWN_MEMBER_ID, this alone is enough to suggest that it's doing a restart. We should expect no rebalance in this case.
```

So, if you do a rolling bounce AND add topics, the new topics will not be consumed. You basically have to wait for the group to be joined, and then force an additional join with `ForceRejoin`. Or, you could wait for the group to be joined _and then_ add the topics. It's odd. I don't know why they did this. I'll look a bit more.

Reply:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-814
This is exactly your problem

Reply:
https://issues.apache.org/jira/browse/KAFKA-13435?focusedCommentId=17620653&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17620653

It's still not even solved with KIP-814.

Reply:
I'm going to merge a similar patch to this into franz-go. Although this isn't technically sound and not defined in any KIP, this is the implementation within Kafka and within Redpanda. If the JoinResponse version is <= 8, then we can be _somewhat_ sure that if the member ID is prefixed with the instance ID, then this is the leader. However, just below (line 1112 your patch), I will also add skipping the assignment. Brokers actually do not reply with member protocol data, so this cannot actually do a balance right away.

Reply:
I can also work around KIP-814 limitations within the client, but your brokers must be v3.2+

Reply:
That's awesome, thanks!

Reply:
See this PR: https://github.com/twmb/franz-go/pull/229

Reply:
@twmb so fast, appreciate it!

Reply:
I'll be releasing this tonight, just a little bit more to go

Reply:
(released)